### PR TITLE
ChangeAndIfToEarlyReturnRector: added a failing testcase for continue-in-foreach

### PR DESCRIPTION
**Pull request by clxmstaab**
refs https://github.com/rectorphp/rector/issues/4693

this failing test just represents a example. same is true for `for` and `while` loops

### CODE REVIEW

**Comment by TomasVotruba** [comment]
Thanks :+1: 

Just curious, why do you have 2 ids on Github? Is that on purpose or a different person? :)

**Comment by clxmstaab** [comment]
> Just curious, why do you have 2 ids on Github? Is that on purpose or a different person? :)

one personal account and one corporate account :)

**Comment by TomasVotruba** [comment]
Ah :). Why you use both of them in Rector? 
It very confusing for me and hard to track what I wrote to whom :D

**Comment by clxmstaab** [comment]
> Ah :). Why you use both of them in Rector?
> It very confusing for me and hard to track what I wrote to whom :D

hehe sorry. depends on which browser I surf.

from the company I only have commit access with clxmstaab.

**Comment by TomasVotruba** [comment]
Allright :) 

Would you like to fix this one? It's a good first issue

**Comment by clxmstaab** [comment]
> Would you like to fix this one? It's a good first issue

I will give it a try.

having a look at the code, I found out:
- I need to adjust this line https://github.com/rectorphp/rector/blob/080e5cbe00b30073395091d162e8934ca29ce261/rules/solid/src/Rector/If_/ChangeAndIfToEarlyReturnRector.php#L207
- I need to check whether the current if has a foreach/while/for loop as one of its parents. If so, use "continue" otherwise use a "return" as we have right now

now I need to figure out, how to detect whether the current if is within a loop

**Comment by clxmstaab** [comment]
I think I got it right for my new testcase. the change regresses a already existing test though.

I am pretty sure though that the test-expectations of `if_in_foreach.php.inc` is wrong.
the expected result changes the meaning of the original code, because it no longer guarantees that the loop will only process the first element

**Comment by clxmstaab** [comment]
IMO `if_in_foreach` should be adjusted like

```
Index: rules/solid/tests/Rector/If_/ChangeAndIfToEarlyReturnRector/Fixture/if_in_foreach.php.inc
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/rules/solid/tests/Rector/If_/ChangeAndIfToEarlyReturnRector/Fixture/if_in_foreach.php.inc b/rules/solid/tests/Rector/If_/ChangeAndIfToEarlyReturnRector/Fixture/if_in_foreach.php.inc
--- a/rules/solid/tests/Rector/If_/ChangeAndIfToEarlyReturnRector/Fixture/if_in_foreach.php.inc	(revision ddbe858d5eb66c17410222c62194cfd6f9757a20)
+++ b/rules/solid/tests/Rector/If_/ChangeAndIfToEarlyReturnRector/Fixture/if_in_foreach.php.inc	(date 1606466114298)
@@ -42,6 +42,8 @@
                 return;
             }
             $this->processNameOrIdentifier($node);
+
+            return;
         }
 
         return;
```

but then I still don't know how to identify this case properly to choose correctly between `return` and `continue`

**Comment by TomasVotruba** [comment]
> IMO if_in_foreach should be adjusted like

I've checked and you're right :+1:

**Comment by TomasVotruba** [comment]
> but then I still don't know how to identify this case properly to choose correctly between return and continue.

That could be done. But we'd need a new rule to handle returns and their combination. I think it would be easier to maintain one rule that decides what to use in one place.

File: rules/solid/src/Rector/If_/ChangeAndIfToEarlyReturnRector.php [hunk 1]
```diff
@@ -279,4 +286,18 @@ private function isLastIfOrBeforeLastReturn(If_ $if): bool
         }
         return $nextNode instanceof Return_;
     }
+
+    private function isIfInLoop(If_ $if): bool
+    {
+        $parent = $if->getAttribute(AttributeKey::PARENT_NODE);
```
Comment (TomasVotruba) [inline rules/solid/src/Rector/If_/ChangeAndIfToEarlyReturnRector.php:292]: Here might be used: 

```php
$this->betterNodeFinder->findFirstParentInstanceOf(...)
```

**Comment by clxmstaab** [comment]
> That could be done. But we'd need a new rule to handle returns and their combination. I think it would be easier to maintain one rule that decides what to use in one place.

I agree. but I don't know how/where to place this logic in the rule, without breaking the existing stuff  :-)

**Comment by TomasVotruba** [comment]
Personally, I'd start with 1 fixture only, fix and then merge it. 
It's easier to get the work done, because there is less problems at once.

Which of those 2 fixtures do you find easier to fix?

File: rules/solid/src/Rector/If_/ChangeAndIfToEarlyReturnRector.php (hunk 1 above)
Comment (clxmstaab) [reply rules/solid/src/Rector/If_/ChangeAndIfToEarlyReturnRector.php]: done

**Comment by clxmstaab** [comment]
> Which of those 2 fixtures do you find easier to fix?

the new one is already fixed. it regresses `if_in_foreach` atm.

File: rules/solid/src/Rector/If_/ChangeAndIfToEarlyReturnRector.php [hunk 2]
```diff
@@ -198,13 +201,17 @@ private function getBooleanAndConditions(BooleanAnd $booleanAnd): array
...
      */
-    private function createInvertedIfNodesFromConditions(array $conditions): array
+    private function createInvertedIfNodesFromConditions(bool $isIfInLoop, array $conditions): array
     {
         $ifs = [];
         foreach ($conditions as $condition) {
             $invertedCondition = $this->conditionInverter->createInvertedCondition($condition);
             $if = new If_($invertedCondition);
-            $if->stmts = [new Return_()];
+            if ($isIfInLoop) {
```
Comment (TomasVotruba) [inline rules/solid/src/Rector/If_/ChangeAndIfToEarlyReturnRector.php:210]: I'm looking int to this and the only difference is extra `return` after the `if`.

<br>

So basically here must be check for the `return;` in $foreach->stmts after the `if`.

Something like:

```suggestion
            if ($isIfInLoop || $this->isFollowedByReturn((array) $foreach->stmts) {
```

**Comment by TomasVotruba** [comment]
> the new one I introduced in the first step of this PR is already fixed. but the PR regresses if_in_foreach atm.

I see. Personaly I'd try to fix the regressoin first without adding a new test fixture. It's much faster to work with lower cognitive load. I'm looking into this PR and I'm already exhausted from switching between 2 cases :D.

**Comment by clxmstaab** [comment]
> I'm looking into this PR and I'm already exhausted from switching between 2 cases :D.

ok, let me put the fix of the existing test in a separate PR

**Comment by clxmstaab** [comment]
I feel this should now be fine. 

After https://github.com/rectorphp/rector/pull/4711 was merged, we just need to rebase.

just make sure we merge https://github.com/rectorphp/rector/pull/4711 before, so the diff of this PR gets smaller (it contains #4711 right now)

**Comment by clxmstaab** [comment]
should be good to go now

**Review by TomasVotruba — APPROVED** [review]

**Comment by TomasVotruba** [comment]
Thank you, very good job :+1:

**Pull request merged**